edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -305,71 +310,137 @@ void TriggerReconnect(object sender, EventArgs e)
+            var upstreamTask = Task.Factory.StartNew(this.UpstreamLoop, TaskCreationOptions.LongRunning);
+
+            return Task.WhenAll(downstreamTask, upstreamTask);
+        }
+
+        async Task StopForwardingLoopsAsync()
+        {
+            this.downstreamPublications.ForEach(channel => channel.Writer.Complete());
+            this.upstreamPublications.ForEach(channel => channel.Writer.Complete());
+
+            await this.forwardingLoops.ForEachAsync(loop => loop);
+
+            this.forwardingLoops = Option.None<Task>();
+            this.downstreamPublications = Option.None<Channel<MqttPublishInfo>>();
+            this.upstreamPublications = Option.None<Channel<MqttPublishInfo>>();
+        }
+
+        void CreateMessageChannels()
+        {
+            this.downstreamPublications = Option.Some(Channel.CreateUnbounded<MqttPublishInfo>(
+                                            new UnboundedChannelOptions
                                             {
-                                                accepted = await consumer.HandleAsync(publishInfo);
-                                                if (accepted)
-                                                {
-                                                    Events.MessageForwarded(consumer.GetType().Name, accepted, publishInfo.Topic, publishInfo.Payload.Length);
-                                                    break;
-                                                }
-                                            }
-                                            catch (Exception e)
+                                                SingleReader = true,
+                                                SingleWriter = true
+                                            }));
+
+            this.upstreamPublications = Option.Some(Channel.CreateUnbounded<MqttPublishInfo>(
+                                            new UnboundedChannelOptions
                                             {
-                                                Events.FailedToForward(e);
-                                                // Keep going with other consumers...
-                                            }
-                                        }
+                                                SingleReader = true,
+                                                SingleWriter = true
+                                            }));
+        }
 
-                                        if (!accepted)
-                                        {
-                                            Events.MessageNotForwarded(publishInfo.Topic, publishInfo.Payload.Length);
-                                        }
-                                    }
+        async Task DownstreamLoop()
+        {
+            Events.DownstreamForwardingLoopStarted();
+            while (await this.downstreamPublications.Expect(ChannelIsBroken).Reader.WaitToReadAsync())
```

Comment:
nit: extract var `Channel<MqttPublishInfo> channel = this.downstreamPublications.Expect(ChannelIsBroken)`

Reply:
extracted